packages/embedder/embedder.ts

```diff
@@ -0,0 +1,250 @@
+  while (true) {
+    const maybeRedisQItem = await redisClient.zpopmax('embedder:queue', 1)
+    if (maybeRedisQItem.length < 2) return // Q is empty, all done!
+
+    const [id, _] = maybeRedisQItem
+    if (!id) {
+      console.warn(`De-queued undefined item from embedder:queue`)
+      continue
+    }
+    const jobQueueId = parseInt(id, 10)
+    const jobQueueItem = await selectJobQueueItemById(jobQueueId)
+    if (!jobQueueItem) {
+      console.warn(`Unable to fetch EmbeddingsJobQueue.id = ${id}`)
+      continue
+    }
+
+    const metadata = await selectMetadataByJobQueueId(jobQueueId)
+    if (!metadata) {
+      await updateJobState(jobQueueId, 'failed', {
+        stateMessage: `unable to fetch metadata by EmbeddingsJobQueue.id = ${id}`
+      })
+      continue
+    }
+
+    let fullText = metadata?.embedText
+    try {
+      if (!fullText) {
+        fullText = await createEmbeddingTextFrom(jobQueueItem, dataLoader)
+      }
+    } catch (e) {
+      await updateJobState(jobQueueId, 'failed', {
+        stateMessage: `unable to create embedding text: ${e}`
+      })
+      continue
+    }
+
+    const wordCount = countWords(fullText)
+    for (const embeddingModel of modelManager.getEmbeddingModelsIter()) {
+      let embedText = fullText
+      const {maxInputTokens} = embeddingModel.getModelParams()
+      // we're using word count as an appoximation of tokens
+      if (wordCount * WORD_COUNT_TO_TOKEN_RATIO > maxInputTokens) {
+        try {
+          const generator = modelManager.getFirstGenerator()
+          if (!generator) throw new Error(`Generator unavailable`)
+          const textToSummarize = fullText.slice(0, generator.getModelParams().maxInputTokens)
+          embedText = await generator.summarize(textToSummarize, 0.8, maxInputTokens)
+        } catch (e) {
+          await updateJobState(jobQueueId, 'failed', {
+            stateMessage: `unable to summarize long embed text: ${e}`
+          })
+          continue
+        }
+      }
+      // console.log(`embedText: ${embedText}`)
+
+      let embeddingVector: number[]
+      try {
+        embeddingVector = await embeddingModel.getEmbedding(embedText)
+      } catch (e) {
+        await updateJobState(jobQueueId, 'failed', {
+          stateMessage: `unable to get embeddings: ${e}`
+        })
+        continue
+      }
+
+      // complete job, do the following atomically
+      // (1) update EmbeddingsMetadata to reflect model completion
+      // (2) upsert model table row with embedding
+      // (3) delete EmbeddingsJobQueue row
+      const modelTable = embeddingModel.getTableName()
+      await completeJobTxn(modelTable, jobQueueId, metadata, fullText, embedText, embeddingVector)
+      console.log(
+        `embedder: completed ${jobQueueItem.objectType}:${jobQueueItem.refId} -> ${modelTable}`
+      )
+    }
+  }
+}
+
+const tick = async (modelManager: ModelManager) => {
+  console.log(`embedder: tick`)
+  await redlock
```

Comment:
Using a mutex lock allows us to run multiple, parallel embedder processes in a stateless fashion. The critical section of the embedder process ensures that only one embedder process is scraping our DB at a time for historical objects. This is particularly handy for rethinkdb, as it has no native locking constructs.

An alternative would be to define another table in postgres where each row contained something like:

   - objectType
   - timeStart
   - timeStop

And each embedder process would grab a time range with a row lock and begin indexing.

In the end, I decided to achieve parallel safety using the redis mutext lock and a queue as it was simpler to implement than needing to manage Postgres table/row locking.